File: 16.最接近的三数之和.cpp
/*
 * @lc app=leetcode.cn id=16 lang=cpp
 *
 * [16] 最接近的三数之和
 *
 * https://leetcode-cn.com/problems/3sum-closest/description/
 *
 * algorithms
 * Medium (42.47%)
 * Likes:    358
 * Dislikes: 0
 * Total Accepted:    72.7K
 * Total Submissions: 169.5K
 * Testcase Example:  '[-1,2,1,-4]\n1'
 *
 * 给定一个包括 n 个整数的数组 nums 和 一个目标值 target。找出 nums 中的三个整数，使得它们的和与 target
 * 最接近。返回这三个数的和。假定每组输入只存在唯一答案。
 * 
 * 例如，给定数组 nums = [-1，2，1，-4], 和 target = 1.
 * 
 * 与 target 最接近的三个数的和为 2. (-1 + 2 + 1 = 2).
 * 
 * 
 */

// @lc code=start
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int len = nums.size();
        if(len < 3)
            return 0;
        sort(nums.begin(), nums.end());
        int i, low, high;
        int delta = 0x7fffffff, sum, ans;
        for(i = 0; i < len - 2; i++){
            low = i + 1;
            high = len - 1;
            while(low < high){
                sum = nums[i] + nums[low] + nums[high];
                if(sum == target){
                    return sum;
                }
                else if(sum < target){
                    if(target - sum < delta){
                        delta = target - sum;
                        ans = sum;
                    }
                    low++;
                }
                else{
                    if(sum - target < delta){
                        delta = sum - target;
                        ans = sum;
                    }
                    high--;
                }
            }
        }
        return ans;
    }
};
// @lc code=end

File: JavaSolution/16.最接近的三数之和.java
import java.util.Arrays;

/*
 * @lc app=leetcode.cn id=16 lang=java
 *
 * [16] 最接近的三数之和
 *
 * https://leetcode-cn.com/problems/3sum-closest/description/
 *
 * algorithms
 * Medium (42.47%)
 * Likes:    358
 * Dislikes: 0
 * Total Accepted:    72.7K
 * Total Submissions: 169.5K
 * Testcase Example:  '[-1,2,1,-4]\n1'
 *
 * 给定一个包括 n 个整数的数组 nums 和 一个目标值 target。找出 nums 中的三个整数，使得它们的和与 target
 * 最接近。返回这三个数的和。假定每组输入只存在唯一答案。
 * 
 * 例如，给定数组 nums = [-1，2，1，-4], 和 target = 1.
 * 
 * 与 target 最接近的三个数的和为 2. (-1 + 2 + 1 = 2).
 * 
 * 
 */

// @lc code=start

// T:O(n^2) S:O(1)
// 先排序，固定首位，次位和尾位双指针夹
// 测靠近用绝对值判定
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int result = nums[0] + nums[1] + nums[2];
        for (int i = 0; i < nums.length - 2; i++) {
            int left = i + 1;
            int right = nums.length - 1;
            while (left != right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (Math. abs(sum - target) < Math.abs(result - target))
                    result = sum;
                if (sum > target) {
                    right--;
                    // 解决nums[right]重复
                    while (left != right && nums[right] == nums[right + 1])
                        right--;
                } else {
                    left++;
                    // 解决nums[left]重复
                    while (left != right && nums[left] == nums[left - 1])
                        left++;
                }
            }
            // 解决nums[i]重复
            while (i < nums.length - 2 && nums[i] == nums[i + 1])
                i++;
        }
        return result;
    }
}
// @lc code=end

